File: ports/fsharp/Pull.fs
module SpriteTool.Pull

open System
open System.Diagnostics
open System.IO

let private usage () =
    printfn
        "Usage: sprite-tool pull <remote-path> <local-path> [sprite-name]\n\
         \n\
         Examples:\n\
         \  sprite-tool pull /home/sprite/file.txt ./file.txt\n\
         \  sprite-tool pull /home/sprite/mydir ./mydir\n\
         \  sprite-tool pull /tmp/file.txt ./file.txt ember-red-hawk"

    exit 1

let run (args: string list) =
    if args.Length < 2 then
        usage ()

    let remotePath = args.[0]
    let localPath = args.[1]

    let spriteName =
        if args.Length > 2 then Some args.[2] else None

    let sArgs = Sprite.spriteArgs spriteName

    // Check if remote is directory or file
    let checkPsi = ProcessStartInfo("sprite")
    checkPsi.UseShellExecute <- false
    checkPsi.RedirectStandardOutput <- true
    checkPsi.RedirectStandardError <- true

    for a in
        Array.append
            [| "exec" |]
            (Array.append sArgs
                [| "bash"
                   "-c"
                   sprintf "[ -d '%s' ] && echo dir || echo file" remotePath |]) do
        checkPsi.ArgumentList.Add a

    use checkProc = Process.Start checkPsi
    let checkOutput = checkProc.StandardOutput.ReadToEnd().Trim()
    checkProc.WaitForExit()

    let isDir = (checkOutput = "dir")

    if isDir then
        printfn "Pulling directory: %s -> %s" remotePath localPath
        Directory.CreateDirectory localPath |> ignore

        // sprite exec [-s name] tar czf - -C remotePath .
        let spritePsi = ProcessStartInfo("sprite")
        spritePsi.UseShellExecute <- false
        spritePsi.RedirectStandardOutput <- true

        for a in
            Array.append
                [| "exec" |]
                (Array.append sArgs [| "tar"; "czf"; "-"; "-C"; remotePath; "." |]) do
            spritePsi.ArgumentList.Add a

        use spriteProc = Process.Start spritePsi

        // tar xzf - -C localPath
        let tarPsi = ProcessStartInfo("tar")
        tarPsi.UseShellExecute <- false
        tarPsi.RedirectStandardInput <- true

        for a in [| "xzf"; "-"; "-C"; localPath |] do
            tarPsi.ArgumentList.Add a

        use tarProc = Process.Start tarPsi

        spriteProc.StandardOutput.BaseStream.CopyTo(tarProc.StandardInput.BaseStream)
        tarProc.StandardInput.Close()
        tarProc.WaitForExit()
        spriteProc.WaitForExit()

        if tarProc.ExitCode <> 0 then
            eprintfn "Error: pull failed"
            exit 1
    else
        printfn "Pulling file: %s -> %s" remotePath localPath

        let localDir = Path.GetDirectoryName(Path.GetFullPath localPath)

        if not (String.IsNullOrEmpty localDir) then
            Directory.CreateDirectory localDir |> ignore

        // sprite exec [-s name] cat remotePath > localPath
        let spritePsi = ProcessStartInfo("sprite")
        spritePsi.UseShellExecute <- false
        spritePsi.RedirectStandardOutput <- true

        for a in
            Array.append
                [| "exec" |]
                (Array.append sArgs [| "cat"; remotePath |]) do
            spritePsi.ArgumentList.Add a

        use spriteProc = Process.Start spritePsi

        use fileStream = File.Create localPath
        spriteProc.StandardOutput.BaseStream.CopyTo(fileStream)
        spriteProc.WaitForExit()

        if spriteProc.ExitCode <> 0 then
            eprintfn "Error: pull failed"
            exit 1

    printfn "Done."

File: ports/ocaml/lib/pull.ml
(* pull.ml — sprite-tool pull: pull a file or directory from a sprite *)

let usage () =
  Printf.printf
{|Usage: sprite-tool pull <remote-path> <local-path> [sprite-name]

Examples:
  sprite-tool pull /home/sprite/file.txt ./file.txt
  sprite-tool pull /home/sprite/mydir ./mydir
  sprite-tool pull /tmp/file.txt ./file.txt ember-red-hawk
|};
  exit 1

let resolve_sprite_name args =
  match args with
  | _ :: _ :: name :: _ -> name
  | _ ->
    (match Sys.getenv_opt "SPRITE_NAME" with
     | Some name when String.length (String.trim name) > 0 -> String.trim name
     | _ ->
       Printf.eprintf "Error: sprite name not provided and SPRITE_NAME not set\n%!";
       exit 1)

let run args =
  match args with
  | [] | [_] -> usage ()
  | _ ->
    let remote_path = List.nth args 0 in
    let local_path = List.nth args 1 in
    let sprite_name = resolve_sprite_name args in
    let dry_run = false in

    (* Check if remote path is a directory *)
    let is_dir = Sprite.is_remote_dir ~dry_run ~sprite_name remote_path in

    if is_dir then begin
      Printf.printf "Pulling directory: %s -> %s\n%!" remote_path local_path;
      Sprite.pull_dir ~dry_run ~sprite_name remote_path local_path
    end else begin
      Printf.printf "Pulling file: %s -> %s\n%!" remote_path local_path;
      Sprite.pull_file ~dry_run ~sprite_name remote_path local_path
    end;

    Printf.printf "Done.\n%!"
